### PR TITLE
added toggle for verbose error messages

### PR DESCRIPTION
**Pull request by LeowWB**
actual verbose error messages not included as part of this pull request

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -29,10 +43,28 @@ export function runInContext(
   context: Context,
   options: Partial<IOptions> = {}
 ): Promise<Result> {
+  function getFirstLine(theProgram: Program) {
+    const firstLineOfProgram = theProgram.body[0] as ExpressionStatement
+    if (!!firstLineOfProgram) {
+      const firstLineExpression = firstLineOfProgram.expression as Literal
```
Comment (openorclose) [inline src/index.ts]: Similar to above, I think you should explicitly check expression.type === 'Literal'

File: src/index.ts [hunk 2]
```diff
@@ -29,10 +43,28 @@ export function runInContext(
   context: Context,
   options: Partial<IOptions> = {}
 ): Promise<Result> {
+  function getFirstLine(theProgram: Program) {
+    const firstLineOfProgram = theProgram.body[0] as ExpressionStatement
+    if (!!firstLineOfProgram) {
```
Comment (openorclose) [inline src/index.ts]: I think it's better if you explicitly check if the first line exists, then check if the first line type === 'ExpressionStatement'.

File: src/interpreter-errors.ts [hunk 3]
```diff
@@ -107,7 +112,8 @@ export class UndefinedVariable extends RuntimeSourceError {
   }
 
   public elaborate() {
-    return 'TODO'
+    return `Before you can call ${this.name}, you need to declare it as a variable or a constant.
```
Comment (openorclose) [inline src/interpreter-errors.ts]: Call usually refers to a function call. Its usage here could be confusing.

File: src/interpreter-errors.ts [hunk 4]
```diff
@@ -121,7 +127,7 @@ export class UnassignedVariable extends RuntimeSourceError {
   }
 
   public elaborate() {
-    return 'TODO'
+    return `Please assign a value to ${this.name} first. Example:\n\t${this.name} = 42;`
```
Comment (openorclose) [inline src/interpreter-errors.ts]: This is wrong. Example of when this error occurs.

```javascript

const a = 1;
const f = () => {
  const i_think_i_can_access_outer_a_here = a; // UnassignedVariableError occurs here
  // other stuff;
  const a = 'another use of a';
}
f();
```

This is due to the proper scoping of variables, so access to any variables that are ever declared in the current scope is blocked until the statement where it is being declared is evaluated.

File: src/interpreter-errors.ts [hunk 5]
```diff
@@ -1,4 +1,9 @@
-/* tslint:disable:max-classes-per-file */
+/*
```
Comment (openorclose) [inline src/interpreter-errors.ts]: Not true. InvalidNumberOfArguments is not a JS error but a Source one for example. Should this mean that we should try to separate these away from the JS ones?

File: src/rules/noDeclareMutable.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import astr = require('astring')
```
Comment (openorclose) [inline src/rules/noDeclareMutable.ts]: Why is this require instead of import?

**Comment by thomastanck** [comment]
You'll have to sync master after #174 is merged and update tests. There is a refactor to the test API.

File: src/repl/repl.ts [hunk 7]
```diff
@@ -1,24 +1,30 @@
...
+      }
     }
-  })
+  )
 }
 
 function main() {
-  const chapter = process.argv.length > 2 ? parseInt(process.argv[2], 10) : 1
+  // TODO const chapter alternative case should be 1
+  const chapter = process.argv.length > 2 ? parseInt(process.argv[2], 10) : 4
```
Comment (thomastanck) [inline src/repl/repl.ts]: Did you forget to undo this change or was this intended?

File: src/repl/repl.ts [hunk 8]
```diff
@@ -1,24 +1,30 @@
-import repl = require('repl')
```
Comment (thomastanck) [inline src/repl/repl.ts]: Try not to make refactors in the same PR as a feature PR. It adds additional noise to the PR which reviewers would have to filter out.

File: src/types.ts [hunk 9]
```diff
@@ -1,4 +1,11 @@
+/*
+	This file contains definitions of some interfaces and classes that are used in Source (such as
+	error-related classes).
+*/
+
+/* tslint:disable:interface-functionName max-classes-per-file */
```
Comment (thomastanck) [inline src/types.ts]: It's slightly strange that you had to disable this lint for this file even though it appears that this file didn't change. Was this line added because you made a change in this file and later reverted it?

File: src/index.ts [hunk 10]
```diff
@@ -1,3 +1,4 @@
+import { ExpressionStatement, Literal, Program } from 'estree'
```
Comment (thomastanck) [inline src/index.ts]: Can we have some test cases to check that the toggle works?

I have a hunch that it doesn't. Line 47 may want to check for `'Directive'`. https://github.com/estree/estree/blob/master/es5.md#directive

File: src/index.ts [hunk 11]
```diff
@@ -29,10 +43,32 @@ export function runInContext(
...
 ): Promise<Result> {
+  function getFirstLine(theProgram: Program) {
+    if (theProgram.body[0] && theProgram.body[0].type === 'ExpressionStatement') {
+      const firstLineOfProgram = theProgram.body[0] as ExpressionStatement
+      if (firstLineOfProgram.expression.type === 'Literal') {
+        const firstLineExpression = firstLineOfProgram.expression as Literal
+        if (!!firstLineExpression) {
+          return firstLineExpression.value
+        } else {
+          return undefined
```
Comment (thomastanck) [inline src/index.ts]: You can omit all the `return undefined` here as this function is small and it's obvious that the function returns quickly if the conditions are not met.

File: src/types.ts (hunk 9 above)
Comment (LeowWB) [reply src/types.ts]: You're about right. I added the comment block at the top to help myself understand what was going on when I was still figuring things out on my own. During a merge I decided to keep that block there, while also accepting incoming changes (the comment that disabled the linter). At the time I figured there'd be no harm in leaving a one-line comment in the code. That said, what would you suggest now? Should I remove line 6?

Comment (LeowWB) [reply src/types.ts]: You're about right. I added the comment block at the top to help myself understand what was going on when I was still figuring things out on my own. During a merge I decided to keep that block there, while also accepting incoming changes (the comment that disabled the linter). At the time I figured there'd be no harm in leaving a one-line comment in the code. That said, what would you suggest now? Should I remove line 6?

Comment (thomastanck) [reply src/types.ts]: Yes. You can view the blame for this file to see where these lines come from. (either view it on github or use `git blame`)

File: src/repl/repl.ts (hunk 7 above)
Comment (LeowWB) [reply src/repl/repl.ts]: Nope, it was an oversight on my part. Fixed it.

File: src/types.ts (hunk 9 above)
Comment (LeowWB) [reply src/types.ts]: It seems to have come from #175 ? Either way, I'll remove it.

File: src/index.ts (hunk 10 above)
Comment (LeowWB) [reply src/index.ts]: The toggle works, though I agree it doesn't always work as expected, and there's a nuanced difference in how you would toggle it for interpreter errors and errors that result from breaking Source specs. I have yet to fix that. Meanwhile I'll start pulling over commits from my other branch which does have the test cases. I've also implemented the Directive check, though I had to implement my own interface since the ESTree index that we have doesn't seem to export any such interface.

File: src/repl/repl.ts (hunk 8 above)
Comment (LeowWB) [reply src/repl/repl.ts]: Noted, I'll keep that in mind for future PRs.

File: src/index.ts [hunk 12]
```diff
@@ -35,10 +49,26 @@ export function runInContext(
   context: Context,
   options: Partial<IOptions> = {}
 ): Promise<Result> {
+  function getFirstLine(theProgram: Program) {
+    if (theProgram.body[0] && theProgram.body[0].type === 'ExpressionStatement') {
+      const firstLineOfProgram = theProgram.body[0] as ExpressionStatement
+      const theDirective = (firstLineOfProgram as Directive).directive
+      if (theDirective) {
```
Comment (thomastanck) [inline src/index.ts]: I would check for undefined explicitly. 
maim

File: src/index.ts [hunk 13]
```diff
@@ -35,10 +49,26 @@ export function runInContext(
...
+    }
+
+    return undefined
+  }
   const theOptions: IOptions = { ...DEFAULT_OPTIONS, ...options }
   context.errors = []
   const program = parse(code, context)
   if (program) {
+    if (getFirstLine(program) === 'enable verbose') {
+      verboseErrors = true
```
Comment (thomastanck) [inline src/index.ts]: This kind of `if true then set to true else set to false` pattern is not as good as simply writing `verboseErrors = getFirstLine(program) === 'enable verbose'`

On the other hand, as commented above, might be good to extract this function out for testing

File: src/index.ts [hunk 14]
```diff
@@ -21,11 +22,24 @@ const DEFAULT_OPTIONS: IOptions = {
...
     const line = error.location ? error.location.start.line : '<unknown>'
+    const column = error.location ? error.location.start.column : '<unknown>'
     const explanation = error.explain()
-    return `Line ${line}: ${explanation}`
+    const elaboration = error.elaborate()
+
+    if (verbose) {
+      // TODO currently elaboration is just tagged on to a new line after the error message itself. find a better
+      // way to display it.
+      return `Line ${line}, Column ${column}: ${explanation}. ${elaboration}\n`
```
Comment (thomastanck) [inline src/index.ts]: I think your \n needs to be placed behind the elaboration not after it.

File: src/interpreter-errors.ts [hunk 15]
```diff
@@ -107,7 +107,10 @@ export class UndefinedVariable extends RuntimeSourceError {
   }
 
   public elaborate() {
-    return 'TODO'
+    return `Before you can read the value of ${
+      this.name
+    }, you need to declare it as a variable or a constant.
+		You can do this using the let or const keywords.`
```
Comment (thomastanck) [inline src/interpreter-errors.ts]: Is the second sentence supposed to be on a new line? If so you should get rid of the leading spaces. Spaces are preserved in backtick strings. (the ones surrounded by `)

File: src/interpreter-errors.ts [hunk 16]
```diff
@@ -117,11 +120,17 @@ export class UnassignedVariable extends RuntimeSourceError {
...
 
   public elaborate() {
-    return 'TODO'
+    return `If you're trying to access the value of ${
+      this.name
+    } from an outer scope, please rename the inner ${
+      this.name
+    }. An easy way to avoid this issue in future would be to avoid declaring any variables or constants with the name ${
+      this.name
+    } in the same scope. `
```
Comment (thomastanck) [inline src/interpreter-errors.ts]: Extra space at the end of the string.

File: src/__tests__/interpreter-errors.ts [hunk 17]
```diff
@@ -13,6 +13,13 @@ test('Undefined variable error is thrown', () => {
   `).toMatchInlineSnapshot(`"Line 1: Name im_undefined not declared"`)
 })
 
+test('Undefined variable error is thrown - verbose', () => {
+  return expectParsedError(stripIndent`
+	"enable verbose";
+	im_undefined;
+	`).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (LeowWB) [reply src/__tests__/interpreter-errors.ts]: My reason for using snapshots rather than inline snapshots was for readability - I didn't want the JSON content to make the test suite files messy. That said, you raise a good point, so I'm considering changing the inline snapshots to regular snapshots instead. Thoughts?

Also I feel like your last suggestion would be adding unnecessary functionality, since the interpreter has no actual use for such a function other than in testing.

File: src/index.ts (hunk 14 above)
Comment (LeowWB) [reply src/index.ts]: I'd be against removing the '\n', since the intention was to separate the stack trace from the error message itself, to enhance readability. I'll add another '\n' between the explanation and elaboration.

File: src/index.ts (hunk 12 above)
Comment (LeowWB) [reply src/index.ts]: Sorry, could you rephrase? You'd check for undefined on which name?

**Comment by thomastanck** [comment]
Check on `theDirective`

**Comment by openorclose** [comment]
Is there a reason verbose mode is not another option to be passed into runInContext? Currently native already works this way.

**Comment by LeowWB** [comment]
> Is there a reason verbose mode is not another option to be passed into runInContext? Currently native already works this way.

The main reason is that the verbosity of error messages has little to do with how the program is being run, but rather the way the output is displayed to the user. A verbose error mode doesn't result in any change to the number of errors, or any other aspect of what the interpreter does in the process of running the program.

**Comment by martin-henz** [comment]
I think we should keep such annotations out of the student programs. The reason: Project NGED is working on a collaborative editor. One user may want to have verbose error messages, while another one may want to have terse messages. I can foresee those two going back and forth.

Second reason: We want to keep the language js-slang clean. It should be a strict subset of JavaScript.

It's best to add an option to runInContext, as @openorclose suggests.

In order to test this in the frontend, you may want to look at this frontend branch, implementing an auto-run feature: https://github.com/source-academy/cadet-frontend/tree/auto-run-feature

We would for now add a similar button to the frontend. We can later think about user preference settings for such things.

**Comment by thomastanck** [comment]
The decision was made to mimic the `"use strict";` directive quite some time ago, but we need to revisit that decision now that NATIVE is merged in. Perhaps we should merge this in first since I believe the other PR has changes that depends on this one. After both of these are merged, it shouldn't be hard to make a final PR switching to the option method.

File: src/__tests__/interpreter-errors.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (thomastanck) [reply src/__tests__/interpreter-errors.ts]: Snapshot content should be treated as code, and in this case it's better to inline the snapshots for better viewing. Other snapshots are not inlined because of two reasons:

- they contain much much more information, so inlining would really hurt readability, and
- they act more as a regression test.

In this case we probably want to inline because we also want to test the behaviour.

I don't know if all tests currently follow this convention, but if you find any feel free to PR.

I agree with you @ the other suggestion. No need to overcomplicate the code.

[the remark this answers is not in the document]
Comment (thomastanck) [reply src/__tests__/interpreter-errors.ts]: On another note, please tell me when you're ready for review again.

**Comment by martin-henz** [comment]
Sure. No problem to merge the current solution

"enable verbose";

The directive
"use strict"; 
itself is a bit of a trick. Expression statements like this are consistent with JavaScript syntax (and don't have any effect).

As Thomas suggests, let's re-visit this later.

**Comment by martin-henz** [comment]
Looks like we need to resolve a conflict in the test cases; both branches added test cases in the same location.
